libbpfgo/helpers/rwArray.go

```diff
@@ -0,0 +1,76 @@
+package helpers
+
+import (
+	"sync"
+)
+
+// Like RWMutex, RWArray allows for multiple concurrent readers but
+// only a single writer. The writers lock a mutex while the readers
+// are lock free.
+//
+// It is implemented as an array of slots where each slot holds a
+// value (interface{}) and a boolean marker to indicate if it's
+// in use or not. The insertion (Put) performs a linear probe
+// looking for an available slot as indicated by the in-use marker.
+// While probing, it is not touching the value itself, as it's
+// being read without a lock by the readers.
+
+type RWArray struct {
+	slots []slot
+	mux   sync.Mutex
+}
+
+func NewRWArray(capacity uint) RWArray {
+	return RWArray{
+		slots: make([]slot, capacity),
+	}
+}
+
+func (a *RWArray) Put(v interface{}) int {
+	a.mux.Lock()
+	defer a.mux.Unlock()
+
+	limit := len(a.slots)
+
+	for i := 0; i < limit; i++ {
+		if !a.slots[i].used {
+			a.slots[i].value = v
+			a.slots[i].used = true
+			return i
+		}
+	}
+
+	return -1
```

Comment:
Now that you use slice - why not letting go handle the capacity of the slice?
We can have a default capacity (say, 512), and if no space left - just append one more slot to the array, which will double its capacity. WDYT?

Reply:
Unfortunately we can't reallocate since `Get` does not hold a lock.